modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -241,216 +243,126 @@ def groupby_reduce(
-    @wait_computations_if_benchmark_mode
-    def broadcast_apply(cls, axis, apply_func, left, right, other_name="r"):
-        """
-        Broadcast the `right` partitions to `left` and apply `apply_func` function.
-
-        Parameters
-        ----------
-        axis : {0, 1}
-            Axis to apply and broadcast over.
-        apply_func : callable
-            Function to apply.
-        left : np.ndarray
-            NumPy array of left partitions.
-        right : np.ndarray
-            NumPy array of right partitions.
-        other_name : str, default: "r"
-            Name of key-value argument for `apply_func` that
-            is used to pass `right` to `apply_func`.
-
-        Returns
-        -------
-        np.ndarray
-            NumPy array of result partition objects.
-
-        Notes
-        -----
-        This will often be overridden by implementations. It materializes the
-        entire partitions of the right and applies them to the left through `apply`.
-        """
-
-        def map_func(df, *others):
-            other = (
-                pandas.concat(others, axis=axis ^ 1) if len(others) > 1 else others[0]
-            )
-            return apply_func(df, **{other_name: other})
+                return map_func(df, other=other)
 
-        map_func = cls.preprocess_func(map_func)
-        rt_axis_parts = cls.axis_partition(right, axis ^ 1)
-        return np.array(
-            [
+            apply_func = cls.preprocess_func(apply_func)
+            rt_axis_parts = cls.axis_partition(by, axis ^ 1)
+            mapped_partitions = np.array(
                 [
-                    part.apply(
-                        map_func,
-                        *(
-                            rt_axis_parts[col_idx].list_of_blocks
-                            if axis
-                            else rt_axis_parts[row_idx].list_of_blocks
-                        ),
-                    )
-                    for col_idx, part in enumerate(left[row_idx])
+                    [
+                        part.apply(
+                            apply_func,
+                            *(
+                                rt_axis_parts[col_idx].list_of_blocks
+                                if axis
+                                else rt_axis_parts[row_idx].list_of_blocks
+                            ),
+                        )
+                        for col_idx, part in enumerate(partitions[row_idx])
+                    ]
+                    for row_idx in range(len(partitions))
                 ]
-                for row_idx in range(len(left))
-            ]
+            )
+        else:
+            mapped_partitions = cls.map_partitions(partitions, map_func)
+        return cls.map_partitions(
+            mapped_partitions, reduce_func, axis=axis, enumerate_partitions=True
         )
 
     @classmethod
     @wait_computations_if_benchmark_mode
-    def broadcast_axis_partitions(
+    def map_partitions(
```

Comment:
This `broadcast_apply` is actually only used in two places:
- This branch in [`fillna`](https://github.com/noloerino/modin/blob/4f83a591fdb953459ea2c345dcac31ac55a67acc/modin/core/storage_formats/pandas/query_compiler.py#L1938) (I'll double check if this refactor changed any behavior)
- This case in [`groupby`](https://github.com/modin-project/modin/blob/master/modin/core/dataframe/pandas/partitioning/partition_manager.py#L244), which I inlined because it isn't being used elsewhere besides the `fillna` method.

I do agree that it might be prudent to undo this part of the refactor and preserve the distinction between `broadcast_apply` and `broadcast_axis_partitions` and un-inline the groupby partition calls in case we use it elsewhere in the future.

Reply:
`broadcast_apply` is also used in binary operations:
https://github.com/modin-project/modin/blob/6bdf4e11140272ce44ab4f2481032e7bc3d82d0e/modin/core/dataframe/algebra/binary.py#L87

Reply:
Ah, not sure how I missed that. I'll refactor appropriately.